recipe-server/normandy/recipes/models.py

```diff
@@ -373,6 +393,52 @@ def save(self, *args, **kwargs):
         self.implementation_hash = self.compute_implementation_hash()
         super().save(update_fields=['implementation_hash'])
 
+    def validate_arguments(self, arguments, old_arguments=None):
+        """
+        Test if `arguments` follows all action-specific rules.
+
+        If there is a previous version of the arguments, they should
+        be passed as `old_arguments`. Some validation rules depend on
+        the history of the arguments.
+
+        Raises `ValidationError` if any rules are violated.
+        """
+        if self.name == 'preference-experiment':
+            # Make a default dict that always returns a default dict
+            def default():
+                return defaultdict(default)
+            errors = default()
+
+            # Feature branch slugs should be unique within an experiment.
+            branch_slugs = set()
+            branch_values = set()
+            for i, branch in enumerate(arguments['branches']):
+                if branch['slug'] in branch_slugs:
+                    msg = self.errors['duplicate_branch_slug']
+                    errors['branches'][i]['slug'] = msg
+
+                if branch['value'] in branch_values:
+                    msg = self.errors['duplicate_branch_value']
+                    errors['branches'][i]['value'] = msg
+
+                branch_slugs.add(branch['slug'])
+                branch_values.add(branch['value'])
+
+            # Experiment slugs should be unique.
+            experiment_recipes = Recipe.objects.filter(latest_revision__action=self)
+            experiment_slugs = set()
```

Comment:
We never add the slugs that we've seen to this set, so this doesn't actually validate for conflicting slugs. I'm pretty sure I tested this on my first pass and it worked, but I can't explain why it might've worked then.